src/ast/passes/codegen_llvm.cpp

```diff
@@ -768,6 +768,46 @@ void CodegenLLVM::visit(Call &call)
       }
     }
     expr_ = nullptr;
+  } else if (call.func == "has_key") {
+    auto &arg = *call.vargs.at(0);
+    auto &map = static_cast<Map &>(arg);
+    auto [key, scoped_key_deleter] = getMapKey(map, call.vargs.at(1));
+
+    CallInst *lookup = b_.CreateMapLookup(map, key);
+
+    Function *parent = b_.GetInsertBlock()->getParent();
+    BasicBlock *lookup_success_block = BasicBlock::Create(module_->getContext(),
+                                                          "lookup_success",
+                                                          parent);
+    BasicBlock *lookup_failure_block = BasicBlock::Create(module_->getContext(),
+                                                          "lookup_failure",
+                                                          parent);
+    BasicBlock *lookup_merge_block = BasicBlock::Create(module_->getContext(),
+                                                        "lookup_merge",
+                                                        parent);
+
+    Value *lookup_condition = b_.CreateICmpNE(
+        b_.CreateIntCast(lookup, b_.GET_PTR_TY(), true),
+        b_.GetNull(),
+        "lookup_cond");
+    b_.CreateCondBr(lookup_condition,
+                    lookup_success_block,
+                    lookup_failure_block);
+
+    b_.SetInsertPoint(lookup_success_block);
+
+    b_.CreateBr(lookup_merge_block);
+
+    b_.SetInsertPoint(lookup_failure_block);
+
+    b_.CreateBr(lookup_merge_block);
+
+    b_.SetInsertPoint(lookup_merge_block);
+
+    auto phi = b_.CreatePHI(b_.getInt1Ty(), 2, "result");
+    phi->addIncoming(b_.getTrue(), lookup_success_block);
+    phi->addIncoming(b_.getFalse(), lookup_failure_block);
+    expr_ = phi;
```

Comment:
Why don't just do
```suggestion
    expr_ = b_.CreateICmpNE(
        b_.CreateIntCast(lookup, b_.GET_PTR_TY(), true),
        b_.GetNull(),
        "has_key");
```
and skip all the branching?

Reply:
Cause that would be much simpler and straightforward 🤕